# The nature of sequencing data {#sequencing}

The human microbiome project has initiated the large-scale culture-independent analysis of microbial communities, and transcriptome analysis has led to the study of the transcriptional response to many different disease and ecological states.

However, many studies fail to replicate earlier studies even when the same technologies and strategies are used. As just one example, a multitude of studies have examined the link between autism and the human gut microbiota. These have variously implicated x,y,...,and z microbe as being linked to the condition. In a recent high-profile paper by Hsiao et al. [-@Hsiao:2013], \emph{Bacillus fragilis} was suggested to restore the gut microbiome of a mouse autism model from 'autistic-like' to 'normal'. Examination of the dataset shows that the conclusion was likely due to chance alone [@gloor2016s;@Gloor:2016cjm]. While the autism dataset serves as a facile example, the literature are replete with other examples.

## DNA sequencing describes a random sample of the environment

All high throughput gene sequencing datasets share a common origin and it is important to understand the source of the data, and the standard workflow. In the first step, an arbitrarily large population of DNA molecules is randomly sampled from an environment. The random sampling can be direct, in the case of genomics or metagenomics DNA is made directly from the sample. Sampling can be indiredt in the case of RNA-seq where RNA molecules are sampled after they have been converted to DNA via reverse transcription. Regardless, the DNA molecules are typically fragmented. Sampling can also be indirect in the case of tag-sequencing or single cell sequencing. In tag-sequencing, most typically applied to 16S rRNA gene sequencing, a small defined region is amplified by the PCR. In single cell sequencing the molecules from a single cell are fragmented and amplified.

In the second step, an aliquot of the population of DNA fragments are used to make a library by attaching standard sequences that permit the DNA fragments to be bound to the solid phase sequencing chip particular to a particular platform.

In the third step  a sample of the library is loaded onto a sequencing platform. It is important to understand that all sequencing platforms in widespread use contain a fixed number of locations to which the  DNA fragments can bind.

This standardized workflow can be thought of as three random samples from an urn containing a random assortment of DNA fragments. The first random sample is the sample from the environment itself that is used to make the DNA: what is swabbed, and how much is collected relative to the environment. Depending on the environment, the subset of molecules sampled can be a large and complete sample or a small, incomplete and unrepresentative sample. The second random sample is the subset of DNA fragments that are input into the library preparation reaction. Here, the number of fragments available depends on the initial amount of the DNA, the mean fragment size and the requirements of the library preparation protocol. However, all library preparation protocols required more DNA fragments than can be loaded onto the instrument, an example calculation is given below. The third random sample comprise the subset of DNA fragments from the library that are actually sequenced on the machine.

### Instrument capacity

The Ion Torrent and Ion Proton systems have a chip with a predetermined number of pores on the sequencing chip (10 of thousands to 10s of millions, depending on the chip) that can accept a library fragment. No signal is returned from an empty pore, and the signal is rejected if a pore contains two or more different fragments. Sequencing is successful only when the pore is occupied by a single fragment from the library.

The Illumina sequencing instruments attach the DNA fragments to a glass slide and then each fragment is amplified into millions of identical fragments called clusters, which appear as randomly distributed spots under a microscope. Each different Illumina instrument accommodates a characteristic maximum number of clusters, and if two or more clusters  overlap they are rejected by the software.

Thus, regardless of instrument, the technician must apply a precise number of DNA molecules that maximizes the number of fragments on the sequencing instrument without overloading it. It should be obvious that loading a DNA sequencer is akin to filling the squares on a checkerboard where the goal is to have as many checkers as possible, without overlapping the pieces.

### Example calculation of fragment number

The number of fragments after sequencing is determined by the instrument; an Ilumina MiSeq delivers $\sim 20$M fragments whereas an Illumina NextSeq delivers $\sim 400$M and an Illumina HiSeq can deliver $\sim 250$M reads per lane on each of 8 lanes. The commonly used Nextera DNA library kit is optimized to require 50 ng of DNA per sample. Thus, the number of fragments of DNA (or RNA) molecules in the underlying environment, in general,  vastly outnumbers the number of sequence fragments from which the library is made, and the number of fragments in the library  in turn outnumbers the number of fragments from which sequencing data are ultimately derived. We can do a simple back of the envelope calculation for an example metagenomics sequencing run to show this.

Assume that we have a mixture of bacterial species with a mean genome size of 4 Mb. One mole of genomes would have a mass of $2.64 \times 10^9$ grams. A typical bacterial density for a metagenome would be on the order of at least $10^7$ bacteria per ml of sample, so if we take a 1 ml sample, this is $10^7$ genomes, which corresponds to about 44 ng of DNA.

If the DNA concentration after isolation is 1 ng$/ \mu$l, and one $\mu$l of DNA is taken, this corresponds to ($1\times 10^{-9} \mathrm{\ g}) / (2.64 \times 10^9\ \mathrm{g/mole})  \times (6.02 \times 10^{23} \mathrm{\ genomes/mole}) = 228,000$ genomes. The Illumina Nextera XT kit can be used to make a library with this amount of DNA. Recall that the the DNA is fragmented, typically into 500 bp or smaller sizes. Using a fragment size of 500 bp, this corresponds to approximately $1.8 \times 10^9$ DNA fragments.

In the scenario where a single sample is prepared and run only 1% of DNA fragments in the library will be sequenced on the Illumina MiSeq, and only 22% of the DNA fragmments will be sequenced on the Illumina NextSeq. DNA sequencing rarely involves a single sample, but instead samples are 'multiplexed' on the sequencing run by mixing two or more libraries together. When this occurs the samples have a unique tag, or barcode, attached so that the samples can be uniquely identified post sequencing [@Parameswaran:2007aa;@Andersson:2008;@Gloor:2010]. Barcodes can be added by ligation or by incorporation into PCR primers that are used to amplify the library.  Obviously, increasing the number of samples through multiplexing will result in an even smaller proportion of the fragments in each  library being sequenced.

## Sequencing post processing

After sequencing fragments are grouped in some way

### Mapping

Fragments generated from metagenomic or RNA-seq experiments are aligned to reference sequences corresponding to genes, transcripts or genetic intervals (generically genes). The total number of fragments mapping to a given gene is said to be the read count for that gene and the read count for all genes in a system ranges from 0 (no fragments align to that gene) to the total number of fragments in the sample. The output from these types of experiments is a table of read counts per gene per sample.

Expand this to outline methods

### OTU generation

Fragments generated from tag-sequencing are often merged into operational taxonomic units (OTUs) at some predefined percent identity, or tabulated by the number of identical fragments observed (ASUs). The output from these types of experiments is a table of counts per OTU per sample.
